pkg/middleware/auth_proxy.go

```diff
@@ -150,66 +111,37 @@ func initContextWithAuthProxy(ctx *m.ReqContext, orgID int64) bool {
 		return true
 	}
 
-	// Make sure that we cannot share a session between different users!
-	if getRequestUserId(ctx) > 0 && getRequestUserId(ctx) != query.Result.UserId {
-		// remove session
-		if err := ctx.Session.Destory(ctx.Context); err != nil {
-			log.Error(3, "Failed to destroy session. error: %v", err)
-		}
-
-		// initialize a new session
-		if err := ctx.Session.Start(ctx.Context); err != nil {
-			log.Error(3, "Failed to start session. error: %v", err)
-		}
-	}
-
-	ctx.Session.Set(AUTH_PROXY_SESSION_VAR, proxyHeaderValue)
-
 	ctx.SignedInUser = query.Result
 	ctx.IsSignedIn = true
-	ctx.Session.Set(session.SESS_KEY_USERID, ctx.UserId)
 
-	return true
-}
+	expiration := time.Duration(-setting.AuthProxyLdapSyncTtl) * time.Minute
+	key := fmt.Sprintf("auth-proxy-sync-ttl:%s", proxyHeaderValue)
+	value := query.UserId
 
-var syncGrafanaUserWithLdapUser = func(query *m.LoginUserQuery) error {
-	expireEpoch := time.Now().Add(time.Duration(-setting.AuthProxyLdapSyncTtl) * time.Minute).Unix()
-
-	var lastLdapSync int64
-	if lastLdapSyncInSession := query.ReqContext.Session.Get(session.SESS_KEY_LASTLDAPSYNC); lastLdapSyncInSession != nil {
-		lastLdapSync = lastLdapSyncInSession.(int64)
+	if err = store.Set(key, value, expiration); err != nil {
```

Comment:
We should only update the cache key if it was not found in the cache. Otherwise ldap sync will never happen if the user keeps actively using Grafana.